### PR TITLE
Made deploy command more friendly and accept tarballs, URLs, relative paths, etc.

### PR DESCRIPTION
**Pull request by olehermanse**
Enabled a few different ways to use this command.

Inside masterfiles folder:

```
cf-remote deploy --hub hub .
```

Relative path with trailing slash:

```
cf-remote deploy --hub hub ../masterfiles/
```

Tarball directly, without running autotools and tar:

```
cf-remote deploy --hub hub /home/olehermanse/.cfengine/cf-remote/masterfiles.tgz
```

Gzipped tarball from URL:

```
cf-remote deploy --hub hub https://cfengine-package-repos.s3.amazonaws.com/enterprise/Enterprise-3.15.3/misc/cfengine-masterfiles-3.15.3-1.pkg.tar.gz
```

Installed masterfiles folder, without autogen or configure:

```
cf-remote deploy --hub hub /var/cfengine/masterfiles
```

### CODE REVIEW

File: cf_remote/commands.py [hunk 1]
```diff
@@ -475,18 +475,56 @@ def uninstall(hosts):
...
+        urls = [masterfiles]
+        paths = _download_urls(urls)
+        assert len(paths) == 1
+        masterfiles = paths[0]
+        log.debug(f"Deploying downloaded: {masterfiles}")
+    else:
+        masterfiles = os.path.abspath(os.path.expanduser(masterfiles))
+        log.debug(f"Deploy path expanded to: {masterfiles}")
+
+    if masterfiles.endswith("/"):
```
Comment (craigcomstock) [inline cf_remote/commands.py:501]: yes, this is the  bit I wanted changed recently. :+1: but which even the previous business which expands the absolute path might take care of. :)

**Review by craigcomstock — COMMENTED** [review]
looks great. One thing I think might be better though, allowing an "installed" masterfiles to be deployed (autogen not needed because promises.cf already present).

File: cf_remote/commands.py [hunk 2]
```diff
@@ -475,18 +475,56 @@ def uninstall(hosts):
...
+    if not os.path.isdir(masterfiles):
+        log.error(f"'{masterfiles}' must be a directory")
+        return 1
+
+    directory = masterfiles
+
+    if not directory.endswith("/masterfiles"):
+        log.error("The masterfiles directory to deploy must be called 'masterfiles'")
+        return 1
+    if not os.path.isfile(directory + "/autogen.sh"):
```
Comment (craigcomstock) [inline cf_remote/commands.py:516]: Here, I'd like two options: autogen.sh -> promises.cf, or if promises.cf is already present (such as in an "installed" policy set) don't run autogen.sh at all and allow it to proceed.

**Review by nickanderson — COMMENTED** [review]
I think deploying from tar and tar.gz is nice.

does deploying from a directory imply configure and autogen? (craig mentioned it in his comment).

**Comment by olehermanse** [comment]
@nickanderson @craigcomstock The whole reason why I made the command was to ease the process of `git clone + edit + deploy + edit + deploy ...`.

For this use case, I definitely want to run autogen again and again, because I am editing `promises.cf.in` and I want `promises.cf` to be up to date.

The problem with deploying an "installed" directory, i.e. skipping autogen, is that there is no good way of telling whether I should skip autogen. In both cases, it will be a directory with promises.cf inside. I definitely don't want deploy to do some bad detection and end up deploying an outdated promises.cf file if the user didn't want it.

So, in short, I think to address your use-case there should be a new flag `--skip-configure` or similar, to make it explicit what you want.

File: cf_remote/commands.py (hunk 2 above)
Comment (olehermanse) [reply cf_remote/commands.py:516]: But I definitely do want it to run autogen again and again, because I will be editing `promises.cf.in`, and want my `promises.cf` to be updated.

**Comment by nickanderson** [comment]
skip configure makes sense, does it demand to try configuring?

A common case is that people work from the "ready to install tarball" seeding a repo or something with that content.
If there is no .configure or autogen files, perhaps it could skip that when deploying from a directory

File: cf_remote/commands.py (hunk 2 above)
Comment (craigcomstock) [reply cf_remote/commands.py:516]: That's fine actually. If `autogen.sh` is there, run it. If not, let it pass and deploy if `promises.cf` is present since that would be my case of "installed policy".

**Comment by craigcomstock** [comment]
> So, in short, I think to address your use-case there should be a new flag `--skip-configure` or similar, to make it explicit what you want.

As I mentioned in the other comment thread I wouldn't expect `autogen.sh` to be present in an installed policy set, so no need for `--skip-configure` option.

**Comment by olehermanse** [comment]
@craigcomstock That makes sense, if there's no autogen/configure at all, it's easy.

File: cf_remote/commands.py (hunk 2 above)
Comment (olehermanse) [reply cf_remote/commands.py]: Cool, I have added that in the latest commit.

**Review by craigcomstock — APPROVED** [review]
:pizza: :medal_sports:

File: cf_remote/commands.py [hunk 3]
```diff
@@ -476,10 +476,25 @@ def uninstall(hosts):
     return errors
 
 def deploy(hubs, directory):
-    assert(directory.endswith("/masterfiles"))
-    assert(os.path.isfile(directory + "/autogen.sh"))
+    directory = os.path.abspath(os.path.expanduser(directory))
+    log.debug(f"Deploy directory expanded to: {directory}")
+    if directory.endswith("/"):
+        directory = directory[0:-1]
```
Comment (vpodzime) [inline cf_remote/commands.py]: `directory.rstrip("/")` could replace the two lines above

File: cf_remote/commands.py [hunk 4]
```diff
@@ -476,10 +476,25 @@ def uninstall(hosts):
...
+
+    if not directory.endswith("/masterfiles"):
+        log.error("The masterfiles directory to deploy must be called 'masterfiles'")
+        return 1
+    if not os.path.isdir(directory):
+        log.error(f"'{directory}' must be a directory")
+        return 1
+    if not os.path.isfile(directory + "/autogen.sh"):
```
Comment (vpodzime) [inline cf_remote/commands.py]: Could use `os.path.join()`, but I prefer the simplicity of just `+ "/..."`.

**Review by vpodzime — APPROVED** [review]
Looks good to me.

File: cf_remote/ssh.py [hunk 5]
```diff
@@ -57,13 +57,17 @@ def connect_wrapper(host, *args, **kwargs):
...
 
 
-def scp(file, remote, connection=None):
+def scp(file, remote, connection=None, rename=None):
     if not connection:
         with connect(remote) as connection:
-            scp(file, remote, connection)
+            scp(file, remote, connection, rename)
     else:
         print("Copying: '{}' to '{}'".format(file, remote))
```
Comment (larsewi) [inline cf_remote/ssh.py:65]:
```Python
print(f"Copying: '{file}' to '{remote}'")
```
Would make it a tiny bit more neat, from my point of view

File: cf_remote/utils.py [hunk 6]
```diff
@@ -76,7 +76,7 @@ def above_package_path():
 
 
 def is_package_url(string):
-    return bool(re.match("https?://.+/.+\.(rpm|deb|msi)", string))
+    return bool(re.match("https?://.+/.+\.(rpm|deb|msi|tar\.gz|tgz)", string))
```
Comment (larsewi) [inline cf_remote/utils.py:79]: Using raw strings `r""` you can omit the escape characters, which makes the regex a little easier to interpret.

**Review by larsewi — APPROVED** [review]
Looking good !

File: cf_remote/commands.py (hunk 4 above)
Comment (olehermanse) [reply cf_remote/commands.py]: yep

File: cf_remote/ssh.py (hunk 5 above)
Comment (olehermanse) [reply cf_remote/ssh.py:65]: Yep, but IIRC this part was written before we had f-strings :)

File: cf_remote/commands.py (hunk 3 above)
Comment (olehermanse) [reply cf_remote/commands.py]: Not technically the same (rstrip would remove multiple slashes), but I don't think it matters too much here.

File: cf_remote/utils.py (hunk 6 above)
Comment (olehermanse) [reply cf_remote/utils.py:79]: @larsewi good rule in general, but that's not what's happening here. The `\.` is not an escape that python will do anything with, so it will end up in the regex, you don't have to `\\.`. We actually want the `\.` in the regex, meaning specifically dot, not the normal meaning of a dot (wildcard).

In this case, adding the r for raw string makes no difference, there is nothing in that string which python will change;

```
$ python3
Python 3.8.5 (default, Jul 28 2020, 12:59:40)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> "https?://.+/.+\.(rpm|deb|msi|tar\.gz|tgz)"
'https?://.+/.+\\.(rpm|deb|msi|tar\\.gz|tgz)'
>>> r"https?://.+/.+\.(rpm|deb|msi|tar\.gz|tgz)"
'https?://.+/.+\\.(rpm|deb|msi|tar\\.gz|tgz)'
>>> "https?://.+/.+\.(rpm|deb|msi|tar\.gz|tgz)" == r"https?://.+/.+\.(rpm|deb|msi|tar\.gz|tgz)"
True
```

Comment (olehermanse) [reply cf_remote/utils.py:79]: https://github.com/cfengine/cf-remote/pull/8

File: cf_remote/ssh.py (hunk 5 above)
Comment (olehermanse) [reply cf_remote/ssh.py:65]: https://github.com/cfengine/cf-remote/pull/8

File: cf_remote/commands.py (hunk 3 above)
Comment (olehermanse) [reply cf_remote/commands.py]: https://github.com/cfengine/cf-remote/pull/8
